src/Storages/StorageMemory.cpp

```diff
@@ -201,17 +208,83 @@ BlockOutputStreamPtr StorageMemory::write(const ASTPtr & /*query*/, const Storag
 void StorageMemory::drop()
 {
     std::lock_guard lock(mutex);
-    data.clear();
+    data.set(std::make_unique<BlocksList>());
     total_size_bytes.store(0, std::memory_order_relaxed);
     total_size_rows.store(0, std::memory_order_relaxed);
 }
 
+static inline void updateBlockData(Block & old_block, const Block & new_block)
+{
+    for (const auto & it : new_block)
+    {
+        auto col_name = it.name;
+        auto & col_with_type_name = old_block.getByName(col_name);
+        col_with_type_name.column = it.column;
+    }
+}
+
+void StorageMemory::mutate(const MutationCommands & commands, const Context & context)
+{
+    std::lock_guard lock(mutex);
+    auto metadata_snapshot = getInMemoryMetadataPtr();
+    auto storage = getStorageID();
+    auto storage_ptr = DatabaseCatalog::instance().getTable(storage, context);
+    auto interpreter = std::make_unique<MutationsInterpreter>(storage_ptr, metadata_snapshot, commands, context, true);
+    auto in = interpreter->execute();
+
+    in->readPrefix();
+    BlocksList out;
+    Block block;
+    while ((block = in->read()))
+    {
+        out.push_back(block);
+    }
+    in->readSuffix();
+
+    // all column affected
+    if (interpreter->isAffectingAllColumns())
+    {
+        size_t rows = 0;
+        size_t bytes = 0;
+        for (const auto & buffer : out)
+        {
+            rows += buffer.rows();
+            bytes += buffer.bytes();
+        }
+        data.set(std::make_unique<BlocksList>(out));
```

Comment:
Let's move calculation of rows/bytes and assignment out of the branch to avoid code duplication.